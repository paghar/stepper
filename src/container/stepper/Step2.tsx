import React, {useEffect,FC} from "react";
import {useForm} from "react-hook-form";
import {addUsersRequest,updateFromStep} from "../../redux/user/userActions";
import {useSelector, useDispatch} from "react-redux";
import {useRouter} from "next/router";
import {IUser} from "../../data/interface";
import {SpokenLanguage} from "../../data/constants";
import {z} from "zod";
import {zodResolver} from "@hookform/resolvers/zod";

const FormSchema = z.object({
    brithDay:z.string().min(1, {message: "Date can not be empty"}), 
    address:z.string().min(1, {message: "Address can not be empty"}), 
    height:z.number().min(100,{message: "Height cannot subceed 100 centimetres"}).max(230, {message: "Height cannot exceed 230 centimetres"}), 
    weight:z.number().min(45,{message: "Height cannot subceed 45 centimetres"}).max(250, {message: "Weight cannot exceed 250 kilograms"}),   
    spokenLanguage:z.string().array().nullable()
});

type FormSchemaType = z.infer<typeof FormSchema>;


const Step2: FC = () => {

    const usersData = useSelector(state => state) as IUser;
    const dispatch = useDispatch(); 
    const router = useRouter();  

    const {
        register,       
        handleSubmit,
        formState: {errors},
    } = useForm<FormSchemaType>({
        resolver: zodResolver(FormSchema),
    });
   
    const onSubmit = (data:any) => {        
        dispatch(addUsersRequest(data)); 
        router.push({pathname: "/Stepper", query: {formStep:3}});     
    };

    useEffect(()=>{        
        dispatch(updateFromStep(2));
    },[]);   

    const genderItems = SpokenLanguage.map((item,index)=>{        
        return (
            <div className="pl-4" key={index}>
                <input {...register("spokenLanguage")} type="checkbox" value={item}/>
                <label className="text-gray-700 text-sm h-8 m-2">{item}</label>              
            </div>
        );        
    });

    return (
        <form onSubmit={handleSubmit(onSubmit)} className="bg-white shadow-md rounded p-12 m-10">
            <div className="grid grid-cols-2 gap-4">

                <div>
                    <label className="block text-gray-700 text-sm font-bold h-8" >
                        Date of Birth:
                    </label>
                    <input 
                        className="appearance-none border rounded h-16 w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline" 
                        id="brithDay" 
                        type="text"                    
                        {...register("brithDay")}                       
                    />
                    {errors.brithDay && (
                        <p className="text-sm text-red-600 mt-1">{errors.brithDay.message}</p>
                    )}
                </div>

                <div>
                    <label className="block text-gray-700 text-sm font-bold h-8 " >
                        Address:
                    </label>
                    <input 
                        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline" 
                        id="address" 
                        type="text"                    
                        {...register("address")}                       
                    />
                    {errors.address && (
                        <p className="text-sm text-red-600 mt-1">{errors.address.message}</p>
                    )}
                </div>

                <div>
                    <label className="block text-gray-700 text-sm font-bold h-8 " >
                        Height:
                    </label>
                    <input 
                        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline" 
                        id="height" 
                        type="number"    
                        { ...register("height", {valueAsNumber: true} ) }                                            
                    />
                    {errors.height && (
                        <p className="text-sm text-red-600 mt-1">{errors.height.message}</p>
                    )}
                </div>

                <div>
                    <label className="block text-gray-700 text-sm font-bold h-8" >
                        Weight:
                    </label>
                    <input 
                        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline" 
                        id="Weight:" 
                        type="number"  
                        { ...register("weight", {valueAsNumber: true} ) }  
                        defaultValue={usersData.weight}
                    />
                    {errors.weight && (
                        <p className="text-sm text-red-600 mt-1">{errors.weight.message}</p>
                    )}
                </div>

                <div>
                    <label className="block text-gray-700 text-sm font-bold h-8" >                        
                        Spoken Lanquage:                 
                    </label>   
                    {genderItems}             
                </div>
            </div>
 
            <div className="grid grid-cols-8 mt-4">
                <div className="col-span-7"></div>
                <button 
                    type="submit"
                    className="w-full inline-block px-6 py-2 border-2 border-blue-600 text-blue-600 font-medium
                    text-xs leading-normal uppercase rounded hover:bg-blue hover:bg-opacity-5 focus:outline-none 
                    focus:ring-0 transition duration-150 ease-in-out h-16 "
                >
                 next step
                </button>
            </div>     
            
        </form>
    );
};

export default Step2;
