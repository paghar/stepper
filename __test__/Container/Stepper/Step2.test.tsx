import renderer from "react-test-renderer";
import Step2 from "../../../src/container/stepper/Step2";
import {Provider} from "react-redux";

import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore();
const store = mockStore();

jest.mock("next/router", () => ({
    useRouter: jest.fn()
}));

describe(">>>> Stepper", () => {
    test("++++ Snapshot of Stepper2 form", () => {       
        const component = renderer.create(
            <Provider store={store}>
                <Step2/>
            </Provider>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

export {};

