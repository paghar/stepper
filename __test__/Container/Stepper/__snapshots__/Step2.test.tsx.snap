// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>> Stepper ++++ Snapshot of Stepper1 form 1`] = `
<form
  className="bg-white shadow-md rounded p-12 m-10"
  onSubmit={[Function]}
>
  <div
    className="flex flex-row flex-wrap gap-10"
  >
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8"
      >
        Date of Birth:
      </label>
      <input
        className="appearance-none border rounded h-16 w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="brithDay"
        name="brithDay"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8 "
      >
        Address:
      </label>
      <input
        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="address"
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8 "
      >
        Height:
      </label>
      <input
        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="height"
        name="height"
        onBlur={[Function]}
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8"
      >
        Weight:
      </label>
      <input
        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="Weight:"
        name="weight"
        onBlur={[Function]}
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="lg:basis-1/6 md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8"
      >
        Spoken Lanquage:
      </label>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="German"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          German
        </label>
      </div>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="Franch"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          Franch
        </label>
      </div>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="Spanish"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          Spanish
        </label>
      </div>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="Italian"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          Italian
        </label>
      </div>
    </div>
  </div>
  <div
    className="flex flex-row-reverse m-8"
  >
    <button
      className="w-full inline-block px-6 py-2 border-2 border-blue-600 text-blue-600 font-medium
 text-xs leading-normal uppercase rounded hover:bg-blue hover:bg-opacity-5 focus:outline-none 
 focus:ring-0 transition duration-150 ease-in-out h-16 lg:basis-1/6 md:basis-full sm:basis-full basis-full "
      type="submit"
    >
      next step
    </button>
  </div>
</form>
`;

exports[`>>>> Stepper ++++ Snapshot of Stepper2 form 1`] = `
<form
  className="bg-white shadow-md rounded p-12 m-10"
  onSubmit={[Function]}
>
  <div
    className="flex flex-row flex-wrap gap-10"
  >
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8"
      >
        Date of Birth:
      </label>
      <input
        className="appearance-none border rounded h-16 w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="brithDay"
        name="brithDay"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8 "
      >
        Address:
      </label>
      <input
        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="address"
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8 "
      >
        Height:
      </label>
      <input
        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="height"
        name="height"
        onBlur={[Function]}
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="lg:basis-[48%] md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8"
      >
        Weight:
      </label>
      <input
        className="appearance-none border h-16 rounded w-full text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        id="Weight:"
        name="weight"
        onBlur={[Function]}
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="lg:basis-1/6 md:basis-full sm:basis-full basis-full"
    >
      <label
        className="block text-gray-700 text-sm font-bold h-8"
      >
        Spoken Lanquage:
      </label>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="German"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          German
        </label>
      </div>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="Franch"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          Franch
        </label>
      </div>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="Spanish"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          Spanish
        </label>
      </div>
      <div
        className="pl-4"
      >
        <input
          name="spokenLanguage"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value="Italian"
        />
        <label
          className="text-gray-700 text-sm h-8 m-2"
        >
          Italian
        </label>
      </div>
    </div>
  </div>
  <div
    className="flex flex-row-reverse m-8"
  >
    <button
      className="w-full inline-block px-6 py-2 border-2 border-blue-600 text-blue-600 font-medium
 text-xs leading-normal uppercase rounded hover:bg-blue hover:bg-opacity-5 focus:outline-none 
 focus:ring-0 transition duration-150 ease-in-out h-16 lg:basis-1/6 md:basis-full sm:basis-full basis-full "
      type="submit"
    >
      next step
    </button>
  </div>
</form>
`;
